[NEXT-1017] Intercepted routes: error on Vercel builds only

### Verify canary release

* [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: N/A
      pnpm: 8.1.0
    Relevant packages:
      next: 13.3.1-canary.7
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue

[https://intercepted-modals.vercel.app](https://intercepted-modals.vercel.app)

### To Reproduce

Clone and deploy the following project to Vercel: [https://github.com/alex-ketch/intercepted-modals](https://github.com/alex-ketch/intercepted-modals).
An existing deployment can be seen here: [https://intercepted-modals.vercel.app](https://intercepted-modals.vercel.app)

## 404 Error:

* Open [https://intercepted-modals.vercel.app](https://intercepted-modals.vercel.app)
* Click "intercepting-parallel-modal" link
* Click on any of the links listed under "Feed for userId"
* See 404 error
* Refresh the page. Content of photo detail page is present.

## 500 Error:

* Open [https://intercepted-modals.vercel.app](https://intercepted-modals.vercel.app)
* Click "dynamic-sibling-route" link
* Click on any of the links listed under "Feed for userId"
* See 500 error
* Refresh the page. 500 error is still present.

### Describe the Bug

When a project using intercepted routes is deployed to Vercel, a runtime error is encountered when attempting to open the intercepted route. What makes it hard to debug is that this error is not present during local development builds, or `vercel dev` commands.

There are two similar scenarios which both cause errors:

* 404 error for an intercepted route for a different path segment (e.g. from `/intercepting-parallel-modal/userId` intercepting `intercepting-parallel-modal/photo/1`)
* 500 error for intercepted route for a the current path segment (e.g. from `dynamic-sibling-route` intercepting `/dynamic-sibling-route/3`)

The above linked reproduction repo is directly based on the e2e tests found here: [https://github.com/vercel/next.js/tree/canary/test/e2e/app-dir/parallel-routes-and-interception/app/(group)](https://github.com/vercel/next.js/tree/canary/test/e2e/app-dir/parallel-routes-and-interception/app/(group)). I've added another use case of catching a dynamic path for the current path under the `dynamic-sibling-route` folder.

When attempting to visit the modal route a 404 error is shown, but no relevant logs are seen in Vercel.
When attempting to visit a route for a dynamic sibling route the following 500 error is logged.

```
Error: Cannot find module '/var/task/.next/server/app/(group)/dynamic-sibling-route/@modal/(.)/[id]/page.js'
Require stack:
- /var/task/node_modules/next/dist/server/require.js
- /var/task/node_modules/next/dist/server/next-server.js
- /var/task/___next_launcher.cjs
    at Module._resolveFilename (node:internal/modules/cjs/loader:1039:15)
    at mod._resolveFilename (/var/task/node_modules/next/dist/build/webpack/require-hook.js:23:32)
    at Module._load (node:internal/modules/cjs/loader:885:27)
    at Module.require (node:internal/modules/cjs/loader:1105:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.requirePage (/var/task/node_modules/next/dist/server/require.js:134:12)
    at /var/task/node_modules/next/dist/server/load-components.js:51:73
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.loadComponentsImpl [as loadComponents] (/var/task/node_modules/next/dist/server/load-components.js:51:26)
    at async NextNodeServer.findPageComponentsImpl (/var/task/node_modules/next/dist/server/next-server.js:667:36) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/var/task/node_modules/next/dist/server/require.js',
    '/var/task/node_modules/next/dist/server/next-server.js',
    '/var/task/___next_launcher.cjs'
  ],
  page: '/dynamic-sibling-route/2'
}
RequestId: fda2079e-9e18-4eda-aae7-fea0407cb78b Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

---

* [CodeSandbox](https://codesandbox.io/p/github/alex-ketch/intercepted-modals/main?file=%2FREADME.md&workspace=%257B%2522activeFilepath%2522%253A%2522%252FREADME.md%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clgh77bc700153b616e55865c%2522%253A%257B%2522key%2522%253A%2522clgh77bc700153b616e55865c%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A41315%252C%2522key%2522%253A%2522clgh77j1o006v3b61t7ee6iee%2522%252C%2522isMinimized%2522%253Afalse%252C%2522path%2522%253A%2522%252Fdynamic-sibling-route%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clgh77h9b004a3b61g3uqcvqu%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clgh77bc700153b616e55865c%2522%252C%2522spacesOrder%2522%253A%255B%2522clgh77bc700153b616e55865c%2522%255D%252C%2522hideCodeEditor%2522%253Afalse%257D): error is not present
  * incidentally the intercepting-parallel-modal example also demonstrates issue with clearing the slot contents as mentioned in #48124

Thought I'd mentioned you @feedthejim since I noticed you were working on these features recently. Let me know if you need any further info in debugging this.

### Expected Behavior

Should match behaviour seen in local development builds.
I.e. clicking on a link for an item in the feed should show contents of the detail view.

### Which browser are you using? (if relevant)

*No response*

### How are you deploying your application? (if relevant)

Vercel

<sub>From [SyncLinear.com](https://synclinear.com) | [NEXT-1017](https://linear.app/vercel/issue/NEXT-1017/intercepted-routes-error-on-vercel-builds-only)</sub>

--- a/packages/next/src/lib/generate-interception-routes-rewrites.ts
+++ b/packages/next/src/lib/generate-interception-routes-rewrites.ts
@@ -1,6 +1,7 @@
 import { pathToRegexp } from 'next/dist/compiled/path-to-regexp'
 import { NEXT_URL } from '../client/components/app-router-headers'
 import {
+  INTERCEPTION_ROUTE_MARKERS,
   extractInterceptionRouteInformation,
   isInterceptionRouteAppPath,
 } from '../server/future/helpers/interception-routes'
@@ -9,6 +10,31 @@ import { Rewrite } from './load-custom-routes'
 // a function that converts normalised paths (e.g. /foo/[bar]/[baz]) to the format expected by pathToRegexp (e.g. /foo/:bar/:baz)
 function toPathToRegexpPath(path: string): string {
   return path.replace(/\[([^\]]+)\]/g, ':$1')
+}
+
+// for interception routes we don't have access to the dynamic segments from the
+// referrer route so we mark them as noop for the app renderer so that it
+// can retrieve them from the router state later on. This also allows us to
+// compile the route properly with path-to-regexp, otherwise it will throw
+function voidParamsBeforeInterceptionMarker(path: string): string {
+  let newPath = []
+
+  let foundInterceptionMarker = false
+  for (const segment of path.split('/')) {
+    if (
+      INTERCEPTION_ROUTE_MARKERS.find((marker) => segment.startsWith(marker))
+    ) {
+      foundInterceptionMarker = true
+    }
+
+    if (segment.startsWith(':') && !foundInterceptionMarker) {
+      newPath.push('__NEXT_EMPTY_PARAM__')
+    } else {
+      newPath.push(segment)
+    }
+  }
+
+  return newPath.join('/')
 }
 
 export function generateInterceptionRoutesRewrites(
@@ -26,7 +52,9 @@ export function generateInterceptionRoutesRewrites(
       }/(.*)?`
 
       const normalizedInterceptedRoute = toPathToRegexpPath(interceptedRoute)
-      const normalizedAppPath = toPathToRegexpPath(appPath)
+      const normalizedAppPath = voidParamsBeforeInterceptionMarker(
+        toPathToRegexpPath(appPath)
+      )
 
       // pathToRegexp returns a regex that matches the path, but we need to
       // convert it to a string that can be used in a header value

--- a/packages/next/src/shared/lib/router/utils/prepare-destination.ts
+++ b/packages/next/src/shared/lib/router/utils/prepare-destination.ts
@@ -231,23 +231,16 @@ export function prepareDestination(args: {
 
   let newUrl
 
-  // for interception routes we don't have access to the dynamic segments from the
-  // referrer route so we mark them as noop for the app renderer so that it
-  // can retrieve them from the router state later on. This also allows us to
-  // compile the route properly with path-to-regexp, otherwise it will throw
-  // The compiler also thinks that the interception route marker is an unnamed param, hence '0',
+  // The compiler also that the interception route marker is an unnamed param, hence '0',
   // so we need to add it to the params object.
   if (isInterceptionRouteAppPath(destPath)) {
-    main: for (const segment of destPath.split('/')) {
-      for (const marker of INTERCEPTION_ROUTE_MARKERS) {
-        if (segment.startsWith(marker)) {
-          args.params['0'] = marker
-          break main
-        }
-      }
-      if (segment.startsWith(':')) {
-        const param = segment.slice(1)
-        args.params[param] = '__NEXT_EMPTY_PARAM__'
+    for (const segment of destPath.split('/')) {
+      const marker = INTERCEPTION_ROUTE_MARKERS.find((m) =>
+        segment.startsWith(m)
+      )
+      if (marker) {
+        args.params['0'] = marker
+        break
       }
     }
   }

--- a/test/e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts
+++ b/test/e2e/app-dir/parallel-routes-and-interception/parallel-routes-and-interception.test.ts
@@ -5,8 +5,6 @@ createNextDescribe(
   'parallel-routes-and-interception',
   {
     files: __dirname,
-    // TODO: remove after deployment handling is updated
-    skipDeployment: true,
   },
   ({ next }) => {
     describe('parallel routes', () => {

